pkg/binary/envoy/debug/pii_test.go

```diff
@@ -0,0 +1,128 @@
+		" inbound|9080|http|productpage.default.svc.cluster.local - 10.28.2.27:9080 10.28.2.1:0" +
+		" outbound_.9080_._.productpage.default.svc.cluster.local"
+	logWithOneField := `[2019-09-05T17:53:36.908Z]`
+
+	type Out struct {
+		logs []string
+		err  error
+	}
+	// test cases wrapped up in a map where key is the test case name and value is the struct
+	// defining the inputs and expecte outcome
+	tests := map[string]struct {
+		logs        []string
+		format      string
+		containsPII map[string]bool
+		want        Out
+	}{
+		// positive test cases
+		// only return an array containing one valid log with [%START_TIME%] field hashed
+		"simple": {logs: []string{validLog, logWithOneField},
+			format: format, containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{validLog}, err: nil},
+		},
+		"multiple_pii_fields": {logs: []string{validLog, logWithOneField},
+			format: format, containsPII: map[string]bool{"[%START_TIME%]": true,
+				"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%": true, "%RESPONSE_CODE%": true},
+			want: Out{logs: []string{validLog}, err: nil},
+		},
+
+		// negative test cases
+		"empty_logs": {logs: []string{},
+			format: format, containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{}, err: nil},
+		},
+		"all_invalid_logs": {logs: []string{"test", "test1", "test2"},
+			format: format, containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{}, err: nil},
+		},
+		"empty_format_string": {logs: []string{validLog, logWithOneField},
+			format: "", containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{}, err: nil},
+		},
+		"empty_pii_map": {logs: []string{validLog, logWithOneField},
+			format: format, containsPII: map[string]bool{},
+			want: Out{logs: []string{validLog}, err: nil},
+		},
+		"invalid_pii_map": {logs: []string{validLog, logWithOneField},
+			format: format, containsPII: map[string]bool{"test": true, "test2": true},
+			want: Out{logs: []string{validLog}, err: nil},
+		},
+	}
+
+	// defining subtest with t.Run
+	for name, tc := range tests {
+		logs := tc.logs
+		format := tc.format
+		containsPII := tc.containsPII
+		expectedLogs := tc.want.logs
+		expectedErr := tc.want.err
+		t.Run(name, func(t *testing.T) {
+			actualLogs, actualErr := ProcessLogs(logs, format, containsPII)
+
+			// hash the desired fields for each log in wanted logs
+			for i, log := range expectedLogs {
+				fields, ok := shell.Split(log)
+				if !ok {
+					t.Fatalf("error in splitting the log: %s", log)
+				}
+
+				for j, fieldName := range formats {
+					if containsPII[fieldName] {
+						h, err := Hash(fields[j])
+						if err != nil {
+							t.Fatalf("error in hashing expected field")
+						}
+						fields[j] = h
+					}
+				}
+				expectedLogs[i] = shell.Join(fields)
+			}
+			errorDiff := cmp.Diff(expectedErr, actualErr)
+			logsDiff := cmp.Diff(expectedLogs, actualLogs)
```

Comment:
For now use `reflect.DeepEqual`, this does not work as you might expect. In particular, it does not compare private fields of structs. This won't trip you up in this case, but can definitely cause problems down the line (it works for Google because all the structs they're dealing with are protobufs where all of the fields that matter are public; that's not necessarily true for us in our tests). But see my suggested rewrite to avoid this part.